### PR TITLE
speed up for classification's `build_tree`

### PR DESCRIPTION
**Pull request by Eight1911**
This is for issue #59 and is a complete rewrite of the `build_tree` function for classification that speeds up the computation and adds more pre-pruning options including: `min_samples_split`, `min_samples_leaf`, `min_purity_increase`.

You may notice that a tree is created once and cloned again to fit the current Node/Leaf structure. This is because the information needed for optimization needs to be stored at the node level and is not compatible with the Node/Leaf struct.

This PR is also a taste of what’s to come. If it’s accepted then I will be making more PRs of this sort that optimize regression and build stumps.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -3,6 +3,7 @@ os:
   - linux
 julia:
   - 0.5
```
Comment (ranjanan) [inline .travis.yml]: Are you looking to drop 0.5 support with this PR? If so, you could remove it from here.

File: src/classification/main.jl [hunk 2]
```diff
@@ -111,50 +113,52 @@ function build_stump(labels::Vector, features::Matrix, weights=[0];
...
+    r_labels = labels[(!).(left)]
     return Node(id, thresh,
-                Leaf(majority_vote(labels[split]), labels[split]),
-                Leaf(majority_vote(labels[(!).(split)]), labels[(!).(split)]))
+                Leaf(majority_vote(l_labels), l_labels),
+                Leaf(majority_vote(r_labels), r_labels))
 end
 
-function build_tree(labels::Vector, features::Matrix, nsubfeatures=0, maxdepth=-1; rng=Base.GLOBAL_RNG)
+function build_tree(labels::Vector, features::Matrix, nsubfeatures=0, maxdepth=-1,
```
Comment (ranjanan) [inline src/classification/main.jl]: Perhaps make all these parameters kwargs instead?

**Comment by ranjanan** [comment]
@Eight1911 I created [this gist](https://gist.github.com/ranjanan/a2a9e0bf5a32a303f4961fd581ca0f5d) to test this branch and I found that the f score achieved is consistently inferior compared to master. (~0.1 vs ~0.41 on master). What's more is that the f score varies quite dramatically with every run as well.

**Comment by Eight1911** [comment]
@rajanan For some reason, on my machine, the master branch's f-score is `~0.235` instead of `~0.41`. Scikit-learn achieves an f-score of 0.298, but this is due to the fact that its tree defaults to the gini criterion when called without a keyword argument. 

Specifying the criterion with `DecisionTreeClassifier(criterion="entropy")` gives that scikit-learn's implementation also achieves a score that switches between 0.0553 or 0.1044. This is almost identical to the variations seen in the implementation in the merge branch (0.0555 or 0.1044). On the other hand, switching the default criterion in the merge branch to gini achieves exactly the same f-score as that of scikit-learn (i.e., 0.29802).

File: src/classification/main.jl (hunk 2 above)
Comment (Eight1911) [reply src/classification/main.jl]: IMO, this should be done in another pull request.

**Comment by bensadeghi** [comment]
@Eight1911 Thanks for all the work here!! I've been testing with Julia v0.6.2 and I'm seeing a ~10x speed up for certain runs.
The implementation works well for numerical datasets, but there seems to be issues with categorical features. Take the example below:
```julia
download("https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data","adult.data")

adult = readcsv("adult.data");
features = adult[:, 1:14];
labels   = adult[:, 15];

model = build_tree(labels, features);
preds = apply_tree(model,  features);
confusion_matrix(labels, preds)
```
Here, the same dataset is used for training and testing, and so it's expected that a fully grown tree (ie no pre/post pruning) would return a prediction accuracy of around 100%, which is the case for the current master branch. But for the new implementation, the accuracy is around 76%. I've tried fiddling with the ```NodeMeta``` type, changing types from ```Int64``` to ```Any```, but to no help, and I can't find the root cause of the issue. 

Am I missing something? Your thoughts on this?

**Comment by Eight1911** [comment]
@bensadeghi You're right. I've been a little careless on this. The source of this error came from the fact that the trees for my implementation and for the master branch are divided differently. For the master branch, left nodes contain all data points with values **less than or equal to** the node threshold, while trees in my implementation have left nodes that contain data points with value **strictly less than** the node threshold. 

This didn't matter for numerical features since the threshold is the average between the largest and the smallest values in the left and right children and is therefore always strictly greater than every value in the left node and strictly smaller than every value in the right node. But this is quite a big issue for categorical data points where we can't take averages. I've fixed the issue accordingly.

**Comment by bensadeghi** [comment]
@Eight1911 Great catch!! Yes, that'll make quite the difference with categorical data.
I'm merging in this set of commits, thanks for the great work.

I'll work on getting the package working on Julia v0.7, modifying parameters kwargs as suggested by @ranjanan , and updating the SKL API and help docs.

Please do continue with your contributions to the package. It would be great to have more efficient implementations for regression and adaboost.

**Pull request merged**